src/glfw/window.cc

```diff
@@ -0,0 +1,233 @@
+    Window* bound_handle //
+)
+{
+    setWindowHints();
+
+    WindowType* glfwWindow = glfwCreateWindow(
+        config.width,
+        config.height,
+        config.title,
+        config.fullscreen ? glfwGetPrimaryMonitor() : nullptr,
+        nullptr );
+
+    glfwSetWindowUserPointer( glfwWindow, bound_handle );
+    glfwSetFramebufferSizeCallback( glfwWindow, framebufferSizeCallback );
+
+    logGLFWaction(
+        "create window",
+        getHandleAddressString( bound_handle ) //
+    );
+
+    return glfwWindow;
+} // createWindow
+
+void
+Window::destroyWindow( WindowType* glfwWindow )
+{
+    Window* window = getHandle( glfwWindow );
+
+    glfwDestroyWindow( glfwWindow );
+
+    logGLFWaction(
+        "destroy window",
+        getHandleAddressString( window ) //
+    );
+} // destroyWindow
+
+void
+Window::framebufferSizeCallback(
+    WindowType* glfwWindow,
+    int width,
+    int height //
+)
+{
+    Window* window = getHandle( glfwWindow );
+
+    logGLFWaction(
+        "framebuffer resize",
+        getHandleAddressString( window ),
+        "width = " + std::to_string( width ) + ", height = " + std::to_string( height ) //
+    );
+
+    window->m_resize_callback( window, width, height );
+} // framebufferSizeCallback
+
+FramebufferSizeType
+Window::getFramebufferSize() const&
+{
+    int width = 0;
+    int height = 0;
+
+    glfwGetFramebufferSize( m_handle.get(), &width, &height );
+
+    return std::make_pair( width, height );
+} // getFramebufferSize
+
+void
+Window::setWindowed() const&
+{
+    constexpr int xpos = 0;
+    constexpr int ypos = 0;
+
+    glfwSetWindowMonitor(
+        m_handle.get(),
+        nullptr,
+        xpos,
+        ypos,
+        getFramebufferWidth(),
+        getFramebufferHeight(),
+        GLFW_DONT_CARE );
+
+    logGLFWaction(
+        "choose windowed mode",
+        getHandleAddressString( this ) //
+    );
+} // setWindowed
+
+void
+Window::setFullscreen() const&
+{
+    GLFWmonitor* monitor = glfwGetPrimaryMonitor();
+
+    glfwSetWindowMonitor(
+        m_handle.get(),
+        monitor,
+        0,
+        0,
+        getFramebufferWidth(),
+        getFramebufferHeight(),
+        GLFW_DONT_CARE );
+
+    logGLFWaction( "choose fullscreen mode", getHandleAddressString( this ) );
+} // setFullscreen
+
+WindowManager::WindowManager( ErrorCallbackSignature* error_callback )
+{
+    glfwSetErrorCallback( error_callback );
+    glfwInit();
+
+    logGLFWaction( "initialize library" );
+} // WindowManager
+
+WindowManager::~WindowManager()
+{
+    glfwTerminate();
+
+    logGLFWaction( "terminate library" );
+} // ~WindowManager
+
+bool
+WindowManager::isSuitableVersion()
+{
+    int major = 0;
+    int minor = 0;
+
+    glfwGetVersion( &major, &minor, nullptr );
+
+    if ( major > k_min_major )
+    {
+        return true;
+    }
+
+    if ( major == k_min_major && minor >= k_min_minor )
+    {
+        return true;
+    }
+
+    return false;
+} // isSuitableVersion
```

Comment:
This can be rewritten more concisely with the use of a helper struct and starship operator.

```cpp

struct VersionGLFW
{
  int major;
  int minor;
  auto operator<=>() const = default;
};
```

Then this function could look something like:
```cpp
VersionGLFW version;
glfwGetVersion( &version.major, &version.minor, nullptr );
return version >= VersionGLFW{3, 3}; // Hardcoded 3.3 version for brevity
```

Reply:
Agree.